### PR TITLE
Contracts pallet: removal on idle

### LINKED ISSUE #7817: contracts: Move lazy removal to `on_idle`
Due to the fact that offchain election solutions are submitted as an unsigned transaction they can be prevented from inclusion when the `on_initialize` blocks of any block are taking more weight than `AVERAGE_ON_INITIALIZE_RATIO`. For more information see [this discussion](https://github.com/paritytech/substrate/pull/7740#pullrequestreview-559490300).

Lazy deletions are executed inside an `on_initialize` block and are able to break the `AVERAGE_ON_INITIALIZE_RATIO` assumption depending on configuration and load.  To avoid this scenario we should tune down or completely remove the amount of work that is done during `on_initialize` and move it to an upcoming on_idle (https://github.com/paritytech/substrate/issues/4064) hook.

Please note that the issue with offchain election will be resolved by #7319. The move to the new hook is still useful in order to increase throughput during low utilization.

cc @thiolliere @kianenigma

### PR DESCRIPTION
**Pull request by achimcc**
This PR:
- moves the processing of the deletion queue from the `on_initialize` to the `on_idle` hook
- still invokes the processing of the deletion queue on `on_initialize` if the queue length reached the maximal `DeletionQueueDepth`
- adds test `lazy_removal_does_no_run_on_full_queue_and_full_block` for the edge case of full deletion queue and full block and `lazy_removal_on_full_queue_works_on_initialize` for a full deletion queue
- closes #7817

### CODE REVIEW

**Comment by xlc** [comment]
Will you be interested of our [idle scheduler](https://github.com/AcalaNetwork/Acala/blob/master/modules/idle-scheduler/src/lib.rs) which we are using it to [remove EVM contracts](https://github.com/AcalaNetwork/Acala/blob/59b5ba2fba50fee150e0eff39448a6398f4f619c/modules/evm/src/lib.rs#L1943)

File: frame/contracts/src/weights.rs [hunk 1]
```diff
@@ -182,6 +183,9 @@ impl<T: frame_system::Config> WeightInfo for SubstrateWeight<T> {
 			.saturating_add(T::DbWeight::get().reads(1 as Weight))
 			.saturating_add(T::DbWeight::get().writes(1 as Weight))
 	}
+	fn on_idle() -> Weight {
+		Self::on_initialize()
+	}
```
Comment (athei) [inline frame/contracts/src/weights.rs:188]: This file is auto generated when the benchmarks are run. So we can't make such changes here. Maybe just rename the benchmark to `process_deletion_queue_batch`. Then it is general enough to use it for both.

File: frame/contracts/src/lib.rs [hunk 2]
```diff
@@ -29,7 +29,7 @@
 //! functionality. It can be used with other modules that implement accounts based on [`Currency`].
 //! These "smart-contract accounts" have the ability to instantiate smart-contracts and make calls
 //! to other contract and non-contract accounts.
-//!
+
```
Comment (athei) [inline frame/contracts/src/lib.rs:32]: Why this change? By accident?

File: frame/contracts/src/lib.rs [hunk 3]
```diff
@@ -320,15 +320,30 @@ pub mod pallet {
...
-				.saturating_sub(System::<T>::block_weight().total())
-				.min(T::DeletionWeightLimit::get());
-			Storage::<T>::process_deletion_queue_batch(weight_limit)
-				.saturating_add(T::WeightInfo::on_initialize())
+			// We want to process the deletion_queue in the on_idle hook. Only in the case
+			// that the queue length has reached its maximal depth, we process it here
+			let queue_depth = T::DeletionQueueDepth::get() as usize;
+			let queue_len = <DeletionQueue<T>>::decode_len().unwrap_or(0);
+			let queue_full = queue_len >= queue_depth;
+			if queue_full {
```
Comment (athei) [inline frame/contracts/src/lib.rs:334]: I think we can save this binding. Should be easy enough to understand.

```suggestion
			let max_len = T::DeletionQueueDepth::get() as usize;
			let queue_len = <DeletionQueue<T>>::decode_len().unwrap_or(0);
			if queue_len >= max_len {
```

File: frame/contracts/src/lib.rs [hunk 4]
```diff
@@ -320,15 +320,30 @@ pub mod pallet {
...
+	     	    // We do not want to go above the block limit and rather avoid lazy deletion
+		    	// in that case. This should only happen on runtime upgrades.
```
Comment (athei) [inline frame/contracts/src/lib.rs:336]: Indentation. Remember to run `cargo +nightly fmt` in the substrate repo root.

File: frame/contracts/src/tests.rs [hunk 5]
```diff
@@ -1611,13 +1611,33 @@ fn lazy_removal_works() {
 		assert_matches!(child::get(trie, &[99]), Some(42));
 
 		// Run the lazy removal
-		Contracts::on_initialize(Weight::max_value());
+		Contracts::on_idle(Weight::max_value(), Weight::max_value());
```
Comment (athei) [inline frame/contracts/src/tests.rs:1614]: Isn't the first argument the block number and not some `Weight`? I think it was wrong before (passing a block number and thinking it was `Weight`).

File: frame/contracts/src/lib.rs (hunk 3 above)
Comment (achimcc) [reply frame/contracts/src/lib.rs:334]: I agree, its explicit enough.

File: frame/contracts/src/lib.rs (hunk 4 above)
Comment (achimcc) [reply frame/contracts/src/lib.rs:335]: Done

File: frame/contracts/src/lib.rs (hunk 2 above)
Comment (achimcc) [reply frame/contracts/src/lib.rs:32]: Done

File: frame/contracts/src/tests.rs [hunk 6]
```diff
@@ -1820,6 +1840,10 @@ fn lazy_removal_does_no_run_on_full_block() {
 			DispatchClass::Mandatory,
 		);
 
+		// Fill the deletion queue with dummy values, so that on_initialize attempts
+		// to clear the queue
+		Storage::<Test>::fill_queue_with_dummies();
```
[the remark this answers is not in the document]
Comment (achimcc) [reply frame/contracts/src/tests.rs:1796]: I initially thought its just fills up the existing queue. I now simplified the test. It now only fills the queue and the block and checks that no gas is consumed (i.e. no deletion occurs) in this situation when calling the `on_initialize` hook.

File: frame/contracts/src/tests.rs (hunk 5 above)
Comment (achimcc) [reply frame/contracts/src/tests.rs]: I changed the parameter to `System::block_number()`.

File: frame/contracts/src/weights.rs (hunk 1 above)
Comment (achimcc) [reply frame/contracts/src/weights.rs]: Done

File: frame/contracts/src/lib.rs [hunk 7]
```diff
@@ -320,15 +320,28 @@ pub mod pallet {
...
-			// We do not want to go above the block limit and rather avoid lazy deletion
-			// in that case. This should only happen on runtime upgrades.
-			let weight_limit = T::BlockWeights::get()
-				.max_block
-				.saturating_sub(System::<T>::block_weight().total())
-				.min(T::DeletionWeightLimit::get());
-			Storage::<T>::process_deletion_queue_batch(weight_limit)
-				.saturating_add(T::WeightInfo::on_initialize())
+			// We want to process the deletion_queue in the on_idle hook. Only in the case
+			// that the queue length has reached its maximal depth, we process it here
```
Comment (athei) [inline frame/contracts/src/lib.rs:330]:
```suggestion
			// that the queue length has reached its maximal depth, we process it here.
```

File: frame/contracts/src/tests.rs [hunk 8]
```diff
@@ -1611,13 +1612,33 @@ fn lazy_removal_works() {
...
 }
 
+#[test]
+fn lazy_removal_on_full_queue_works_on_initialize() {
+	ExtBuilder::default().existential_deposit(50).build().execute_with(|| {
+		// Fill the deletion queue with dummy values, so that on_initialize attempts
+		// to clear the queue
+		Storage::<Test>::fill_queue_with_dummies();
+
+		// let queue_depth = <Test as Config>::DeletionQueueDepth;
```
Comment (athei) [inline frame/contracts/src/tests.rs:1629]: Please don't commit commented out code.

File: frame/contracts/src/tests.rs [hunk 9]
```diff
@@ -1762,82 +1783,23 @@ fn lazy_removal_partial_remove_works() {
...
+		// to clear the queue
+		Storage::<Test>::fill_queue_with_dummies();
+
 		// Run the lazy removal without any limit so that all keys would be removed if there
 		// had been some weight left in the block.
-		let weight_used = Contracts::on_initialize(Weight::max_value());
-		let base = <<Test as Config>::WeightInfo as WeightInfo>::on_initialize();
+		let weight_used = Contracts::on_initialize(System::block_number());
+		let base = <<Test as Config>::WeightInfo as WeightInfo>::on_process_deletion_queue_batch();
 		assert_eq!(weight_used, base);
```
Comment (athei) [inline frame/contracts/src/tests.rs:1802]: You can add a check here that the queue is still full.

**Review by athei — APPROVED** [review]

File: frame/contracts/src/tests.rs (hunk 9 above)
Comment (achimcc) [reply frame/contracts/src/tests.rs:1801]: Yes, that makes sense! I added the check! However, I realized that the type `DeletionQueueDepth` is defined as `u32` in the `Config` trait implementation for `Test`. Wouldn't the native type be `usize`?

Comment (athei) [reply frame/contracts/src/tests.rs:1801]: We try to avoid `usize` where possible because it is different in native vs wasm. This is why we rather convert from `usize` to `u32` instead of defing it is `usize`.

**Comment by athei** [comment]
Can you please satisfy the CI (running `cargo +nightly fmt`)?

**Comment by athei** [comment]
Just merge master until CI is green.

File: frame/contracts/src/lib.rs [hunk 10]
```diff
@@ -364,15 +364,28 @@ pub mod pallet {
...
+				// We do not want to go above the block limit and rather avoid lazy deletion
+				// in that case. This should only happen on runtime upgrades.
+				let weight_limit = T::BlockWeights::get()
+					.max_block
+					.saturating_sub(System::<T>::block_weight().total())
+					.min(T::DeletionWeightLimit::get());
+				Storage::<T>::process_deletion_queue_batch(weight_limit)
+					.saturating_add(T::WeightInfo::on_process_deletion_queue_batch())
+			} else {
+				T::WeightInfo::on_process_deletion_queue_batch()
```
Comment (agryaznov) [inline frame/contracts/src/lib.rs:387]: Just to clarify: so here we return non-negotiable weight consumed by block, which we assume to be `T::WeightInfo::on_process_deletion_queue_batch()` as `on_idle` hook will run `process_deletion_queue_batch()` anyway. Right? But what if `on_idle` does not happen in this block?

File: frame/contracts/src/benchmarking/mod.rs [hunk 11]
```diff
@@ -203,7 +203,7 @@ benchmarks! {
 	}
 
 	// The base weight without any actual work performed apart from the setup costs.
```
Comment (agryaznov) [inline frame/contracts/src/benchmarking/mod.rs:205]:
```suggestion
	// The base weight consumed on processing contracts deletion queue.
```

File: frame/contracts/src/tests.rs [hunk 12]
```diff
@@ -1759,82 +1779,29 @@ fn lazy_removal_partial_remove_works() {
 }
 
 #[test]
-fn lazy_removal_does_no_run_on_full_block() {
-	let (code, hash) = compile_module::<Test>("self_destruct").unwrap();
+fn lazy_removal_does_no_run_on_full_queue_and_full_block() {
```
Comment (agryaznov) [inline frame/contracts/src/tests.rs:1782]: How about also adding the test case when the deletion queue is not full but the block is full (so that `on_idle` runs but doesn't clear the queue)?

**Review by agryaznov — COMMENTED** [review]
Hey Achim, the PR looks good! Just highlighted a couple of questions.

File: frame/contracts/src/tests.rs (hunk 12 above)
Comment (achimcc) [reply frame/contracts/src/tests.rs:1782]: > How about also adding the test case when the deletion queue is not full but the block is full? (which probably has two sub-cases: 1) `on_idle` just doesn't run at all, and 2) `on_idle` runs but `remaining_weight` is so low that it doesn't remove anything from deletion queue)

Thanks for the hints! Actually the case 2) that `remaining_weight` is to low for deletion is certainly easy to implement! I'm not too sure about case 1), it should be a general feature of the `on_idle` hook on Substrate, that it doesn't get executed when `remaining_weight` is zero. So I'd assume that this case is covered by tests somewhere else in Substrate?

Comment (athei) [reply frame/contracts/src/tests.rs:1782]: > So I'd assume that this case is covered by tests somewhere else in Substrate?

Could be non zero but still too low to remove a single item.

Comment (achimcc) [reply frame/contracts/src/tests.rs:1784]: Yes, but that what I said. I can implement case 2) but not case 1).

I now added the test `lazy_removal_does_no_run_on_low_remaining_weight` which runs `on_idle` with a remaining weight which is too low, then `on_initialize` with a deletion queue which is not full and finally `on_idle` again with a remaining weight which is high enough for deletion.

**Review by athei — APPROVED** [review]

File: frame/contracts/src/tests.rs [hunk 13]
```diff
@@ -1761,7 +1781,34 @@ fn lazy_removal_partial_remove_works() {
...
+		System::register_extra_weight_unchecked(
+			<Test as system::Config>::BlockWeights::get().max_block,
+			DispatchClass::Mandatory,
+		);
+
+		// Fill the deletion queue with dummy values, so that on_initialize attempts
+		// to clear the queue
+		Storage::<Test>::fill_queue_with_dummies();
+
+		// Run the lazy removal without any limit so that all keys would be removed if there
```
Comment (agryaznov) [inline frame/contracts/src/tests.rs:1796]:
```suggestion
		// Check that on_initialize() tries to perform lazy removal but removes nothing
```

File: frame/contracts/src/tests.rs [hunk 14]
```diff
@@ -1761,7 +1781,34 @@ fn lazy_removal_partial_remove_works() {
...
+		// had been some weight left in the block.
```
Comment (agryaznov) [inline frame/contracts/src/tests.rs:1797]:
```suggestion
		//  as no more weight is left for that.
```

**Review by agryaznov — APPROVED** [review]
